File: scripts/analyze_ml_cms.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <openssl/cms.h>
#include <openssl/bio.h>
#include <openssl/x509.h>
#include <openssl/err.h>

void printCertInfo(const char* label, X509* cert) {
    char subjectBuf[512];
    char issuerBuf[512];

    X509_NAME_oneline(X509_get_subject_name(cert), subjectBuf, sizeof(subjectBuf));
    X509_NAME_oneline(X509_get_issuer_name(cert), issuerBuf, sizeof(issuerBuf));

    bool isSelfSigned = (strcmp(subjectBuf, issuerBuf) == 0);

    std::cout << "\n" << label << ":\n";
    std::cout << "  Subject: " << subjectBuf << "\n";
    std::cout << "  Issuer:  " << issuerBuf << "\n";
    std::cout << "  Self-signed: " << (isSelfSigned ? "YES" : "NO") << "\n";

    // Check key usage
    ASN1_BIT_STRING* keyUsage = (ASN1_BIT_STRING*)X509_get_ext_d2i(cert, NID_key_usage, nullptr, nullptr);
    if (keyUsage) {
        std::cout << "  Key Usage: ";
        if (ASN1_BIT_STRING_get_bit(keyUsage, 0)) std::cout << "digitalSignature ";
        if (ASN1_BIT_STRING_get_bit(keyUsage, 1)) std::cout << "nonRepudiation ";
        if (ASN1_BIT_STRING_get_bit(keyUsage, 2)) std::cout << "keyEncipherment ";
        if (ASN1_BIT_STRING_get_bit(keyUsage, 5)) std::cout << "keyCertSign ";
        if (ASN1_BIT_STRING_get_bit(keyUsage, 6)) std::cout << "cRLSign ";
        std::cout << "\n";
        ASN1_BIT_STRING_free(keyUsage);
    }

    // Check basic constraints
    BASIC_CONSTRAINTS* bc = (BASIC_CONSTRAINTS*)X509_get_ext_d2i(cert, NID_basic_constraints, nullptr, nullptr);
    if (bc) {
        std::cout << "  Is CA: " << (bc->ca ? "YES" : "NO") << "\n";
        BASIC_CONSTRAINTS_free(bc);
    }
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <master_list.cms>\n";
        return 1;
    }

    OpenSSL_add_all_algorithms();
    ERR_load_crypto_strings();

    // Read file
    std::ifstream file(argv[1], std::ios::binary);
    if (!file) {
        std::cerr << "Failed to open file: " << argv[1] << "\n";
        return 1;
    }

    std::vector<uint8_t> data((std::istreambuf_iterator<char>(file)),
                              std::istreambuf_iterator<char>());
    file.close();

    // Parse CMS
    BIO* bio = BIO_new_mem_buf(data.data(), data.size());
    CMS_ContentInfo* cms = d2i_CMS_bio(bio, nullptr);
    BIO_free(bio);

    if (!cms) {
        std::cerr << "Failed to parse CMS structure\n";
        return 1;
    }

    std::cout << "=== Master List CMS Analysis ===\n";
    std::cout << "File: " << argv[1] << "\n";
    std::cout << "Size: " << data.size() << " bytes\n";

    // Get SignerInfo certificates
    STACK_OF(CMS_SignerInfo)* signers = CMS_get0_SignerInfos(cms);
    if (signers) {
        int numSigners = sk_CMS_SignerInfo_num(signers);
        std::cout << "\n--- SignerInfo Certificates (" << numSigners << ") ---\n";
        std::cout << "These are the ACTUAL Master List Signer Certificates (MLSC)\n";

        for (int i = 0; i < numSigners; i++) {
            CMS_SignerInfo* si = sk_CMS_SignerInfo_value(signers, i);
            X509* cert = nullptr;

            // Get the signer certificate
            CMS_SignerInfo_get0_algs(si, nullptr, &cert, nullptr, nullptr);

            if (cert) {
                char label[64];
                snprintf(label, sizeof(label), "SignerInfo #%d (MLSC)", i + 1);
                printCertInfo(label, cert);
            }
        }
    }

    // Get certificates from pkiData
    STACK_OF(X509)* certs = CMS_get1_certs(cms);
    if (certs) {
        int numCerts = sk_X509_num(certs);
        std::cout << "\n--- pkiData Certificates (" << numCerts << ") ---\n";
        std::cout << "These are CSCA and Link Certificates\n";

        for (int i = 0; i < numCerts && i < 5; i++) {  // Show first 5
            X509* cert = sk_X509_value(certs, i);
            if (cert) {
                char label[64];
                snprintf(label, sizeof(label), "pkiData #%d", i + 1);
                printCertInfo(label, cert);
            }
        }

        if (numCerts > 5) {
            std::cout << "\n... and " << (numCerts - 5) << " more certificates in pkiData\n";
        }

        sk_X509_pop_free(certs, X509_free);
    }

    CMS_ContentInfo_free(cms);

    std::cout << "\n=== Analysis Complete ===\n";
    return 0;
}

File: scripts/analyze_ml_cms.py
#!/usr/bin/env python3
"""
Analyze ICAO Master List CMS structure to distinguish:
- MLSC (Master List Signer Certificates) in SignerInfo
- CSCA and Link Certificates in pkiData
"""

import sys
from cryptography import x509
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend
from cryptography.x509.oid import ExtensionOID, NameOID
from asn1crypto import cms, core

def print_cert_info(label, cert):
    """Print certificate information"""
    subject = cert.subject.rfc4514_string()
    issuer = cert.issuer.rfc4514_string()
    is_self_signed = (subject == issuer)

    print(f"\n{label}:")
    print(f"  Subject: {subject}")
    print(f"  Issuer:  {issuer}")
    print(f"  Self-signed: {'YES' if is_self_signed else 'NO'}")

    # Check key usage
    try:
        key_usage = cert.extensions.get_extension_for_oid(ExtensionOID.KEY_USAGE).value
        usages = []
        if key_usage.digital_signature: usages.append("digitalSignature")
        if key_usage.key_cert_sign: usages.append("keyCertSign")
        if key_usage.crl_sign: usages.append("cRLSign")
        if usages:
            print(f"  Key Usage: {', '.join(usages)}")
    except x509.ExtensionNotFound:
        pass

    # Check basic constraints
    try:
        basic_constraints = cert.extensions.get_extension_for_oid(ExtensionOID.BASIC_CONSTRAINTS).value
        print(f"  Is CA: {'YES' if basic_constraints.ca else 'NO'}")
    except x509.ExtensionNotFound:
        pass

def analyze_master_list(file_path):
    """Analyze Master List CMS structure"""
    print("=== Master List CMS Analysis ===")
    print(f"File: {file_path}")

    with open(file_path, 'rb') as f:
        cms_data = f.read()

    print(f"Size: {len(cms_data)} bytes")

    # Parse CMS using asn1crypto
    content_info = cms.ContentInfo.load(cms_data)
    signed_data = content_info['content']

    # Get SignerInfo certificates
    signer_infos = signed_data['signer_infos']
    print(f"\n--- SignerInfo Certificates ({len(signer_infos)}) ---")
    print("These are the ACTUAL Master List Signer Certificates (MLSC)")

    # SignerInfo doesn't directly contain the certificate
    # We need to match it from the certificates collection

    # Get all certificates from the CMS
    certificates = signed_data['certificates']
    print(f"\n--- Total Certificates in CMS: {len(certificates)} ---")

    # Parse all certificates
    cert_list = []
    for i, cert_choice in enumerate(certificates):
        cert_der = cert_choice.chosen.dump()
        cert = x509.load_der_x509_certificate(cert_der, default_backend())
        cert_list.append(cert)

    # Analyze SignerInfo to find which certificates are signers
    signer_cert_serials = []
    for i, signer_info in enumerate(signer_infos):
        sid = signer_info['sid']
        if sid.name == 'issuer_and_serial_number':
            serial = sid.chosen['serial_number'].native
            signer_cert_serials.append(serial)
            print(f"\nSignerInfo #{i+1} uses certificate with serial: {hex(serial)}")

    # Match signer certificates
    print("\n--- MLSC (Master List Signer Certificates) ---")
    mlsc_certs = []
    for cert in cert_list:
        if cert.serial_number in signer_cert_serials:
            mlsc_certs.append(cert)
            print_cert_info(f"MLSC (Serial: {hex(cert.serial_number)})", cert)

    # Other certificates are CSCA/Link
    print(f"\n--- CSCA and Link Certificates ({len(cert_list) - len(mlsc_certs)}) ---")
    count = 0
    for cert in cert_list:
        if cert.serial_number not in signer_cert_serials:
            count += 1
            if count <= 5:  # Show first 5
                subject = cert.subject.rfc4514_string()
                issuer = cert.issuer.rfc4514_string()
                is_link = (subject != issuer)
                cert_type = "Link Certificate" if is_link else "Self-signed CSCA"
                print_cert_info(f"pkiData #{count} ({cert_type})", cert)

    if count > 5:
        print(f"\n... and {count - 5} more CSCA/Link certificates")

    print("\n=== Analysis Complete ===")
    print("\n=== Summary ===")
    print(f"Total SignerInfo (MLSC): {len(mlsc_certs)}")
    print(f"Total CSCA/Link Certificates: {len(cert_list) - len(mlsc_certs)}")

    return mlsc_certs, [c for c in cert_list if c.serial_number not in signer_cert_serials]

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python3 analyze_ml_cms.py <master_list.cms>")
        sys.exit(1)

    analyze_master_list(sys.argv[1])
